accounting/src/delta/delta_data_collection/combine.rs

```diff
@@ -0,0 +1,205 @@
+// Copyright (c) 2021-2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::*;
+
+/// Given two deltas, combine them into one delta, this is the basic delta data composability function
+pub fn combine_deltas<T: Clone + PartialEq>(
+    lhs: &DataDelta<T>,
+    rhs: DataDelta<T>,
+) -> Result<Option<DataDelta<T>>, Error> {
+    match (lhs, rhs) {
+        (DataDelta::Create(_), DataDelta::Create(_)) => Err(Error::DeltaDataCreatedMultipleTimes),
+        (DataDelta::Create(_), DataDelta::Modify((_, d))) => Ok(Some(DataDelta::Create(d))),
+        (DataDelta::Create(_), DataDelta::Delete(_)) => {
+            // if lhs had a creation, and we delete, this means nothing is left and there's a net zero to return
+            Ok(None)
+        }
+        (DataDelta::Modify(_), DataDelta::Create(_)) => Err(Error::DeltaDataCreatedMultipleTimes),
+        (DataDelta::Modify((lhs1, lhs2)), DataDelta::Modify((rhs1, rhs2))) => {
+            if lhs1 == &rhs2 && lhs2 == &rhs1 {
+                Ok(None)
+            } else {
+                Ok(Some(DataDelta::Modify((lhs1.clone(), rhs2))))
+            }
+        }
+        (DataDelta::Modify((d, _)), DataDelta::Delete(_)) => Ok(Some(DataDelta::Delete(d.clone()))),
+        (DataDelta::Delete(lhs), DataDelta::Create(rhs)) => {
+            if lhs == &rhs {
+                Ok(None)
+            } else {
+                Ok(Some(DataDelta::Modify((lhs.clone(), rhs))))
+            }
+        }
+        (DataDelta::Delete(_), DataDelta::Modify(_)) => Err(Error::DeltaDataModifyAfterDelete),
+        (DataDelta::Delete(_), DataDelta::Delete(_)) => Err(Error::DeltaDataDeletedMultipleTimes),
```

Comment:
That's a good question, but I'm not sure how important that is to solve. Curious what @TheQuantumPhysicist think.

One way I can think of is to accumulate deltas on merge instead of folding them. Then on some final step (like combining with data), we can apply accumulated deltas in a predefined order. This is pretty straightforward and avoids the complex logic of recovering information from error types. But again I'm not sure if we actually need this.